src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngAsymmetricAlgorithmCore.cs

```diff
@@ -0,0 +1,144 @@
+                _lazyKey = CngKey.Create(algorithm, null, creationParameters);
+            }
+
+            return _lazyKey;
+        }
+
+        public void SetKey(CngKey key)
+        {
+            Debug.Assert(key != null);
+
+            // If we already have a key, clear it out.
+            if (_lazyKey != null)
+            {
+                _lazyKey.Dispose();
+            }
+
+            _lazyKey = key;
+        }
+
+        public static byte[] HashData(byte[] data, int offset, int count, HashAlgorithmName hashAlgorithm)
+        {
+            // we're sealed and the base should have checked this already.
+            Debug.Assert(data != null);
+            Debug.Assert(offset >= 0 && offset <= data.Length);
+            Debug.Assert(count >= 0 && count <= data.Length);
+            Debug.Assert(!string.IsNullOrEmpty(hashAlgorithm.Name));
+
+            using (HashProviderCng hashProvider = new HashProviderCng(hashAlgorithm.Name, null))
+            {
+                hashProvider.AppendHashData(data, offset, count);
+                byte[] hash = hashProvider.FinalizeHashAndReset();
+                return hash;
+            }
+        }
+
+        public static byte[] HashData(Stream data, HashAlgorithmName hashAlgorithm)
+        {
+            // We're sealed and the base should have checked these already.
+            Debug.Assert(data != null);
+            Debug.Assert(!string.IsNullOrEmpty(hashAlgorithm.Name));
+
+            using (HashProviderCng hashProvider = new HashProviderCng(hashAlgorithm.Name, null))
+            {
+                // Default the buffer size to 4K.
+                byte[] buffer = new byte[4096];
+                int bytesRead;
+                while ((bytesRead = data.Read(buffer, 0, buffer.Length)) > 0)
+                {
+                    hashProvider.AppendHashData(buffer, 0, bytesRead);
+                }
+                byte[] hash = hashProvider.FinalizeHashAndReset();
+                return hash;
+            }
+        }
+
+        public static unsafe byte[] SignHash(CngKey key, byte[] hash, AsymmetricPaddingMode paddingMode, void* pPaddingInfo, int estimatedSize)
+        {
+#if DEBUG
+            estimatedSize = 2;  // Make sure the NTE_BUFFER_TOO_SMALL scenario gets exercised.
+#endif
+            SafeNCryptKeyHandle keyHandle = key.Handle;
+
+            byte[] signature = new byte[estimatedSize];
+            int numBytesNeeded;
+            ErrorCode errorCode = Interop.NCrypt.NCryptSignHash(keyHandle, pPaddingInfo, hash, hash.Length, signature, signature.Length, out numBytesNeeded, paddingMode);
+            if (errorCode == ErrorCode.NTE_BUFFER_TOO_SMALL)
+            {
+                signature = new byte[numBytesNeeded];
+                errorCode = Interop.NCrypt.NCryptSignHash(keyHandle, pPaddingInfo, hash, hash.Length, signature, signature.Length, out numBytesNeeded, paddingMode);
+            }
+            if (errorCode != ErrorCode.ERROR_SUCCESS)
```

Comment:
This if touching the previous if made me read it as an else if (despite the lack of the word else).  Can you separate it out with a blank line? (I'm also a fan of explicit bracing outside of precondition checks, but I'll be happy with the blank line)